azure-eventhubs/src/main/java/com/microsoft/azure/eventhubs/impl/WebSocketProxyConnectionHandler.java

```diff
@@ -0,0 +1,194 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.net.Authenticator;
+import java.net.InetSocketAddress;
+import java.net.PasswordAuthentication;
+import java.net.Proxy;
+import java.net.ProxySelector;
+import java.net.URI;
+import java.util.Base64;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class WebSocketProxyConnectionHandler extends WebSocketConnectionHandler {
+    private static final Logger TRACE_LOGGER = LoggerFactory.getLogger(WebSocketProxyConnectionHandler.class);
+    private final String proxySelectorModifiedError = "ProxySelector has been modified.";
+
+    public static Boolean shouldUseProxy(final String hostName) {
+        final URI uri = createURIFromHostNamePort(hostName, ClientConstants.HTTPS_PORT);
+        final ProxySelector proxySelector = ProxySelector.getDefault();
+        if (proxySelector == null) {
+            return false;
+        }
+
+        final List<Proxy> proxies = proxySelector.select(uri);
+        return isProxyAddressLegal(proxies);
+    }
+
+    public WebSocketProxyConnectionHandler(AmqpConnection messagingFactory) {
+        super(messagingFactory);
+    }
+
+    @Override
+    protected void addTransportLayers(final Event event, final TransportInternal transport) {
+        super.addTransportLayers(event, transport);
+
+        final ProxyImpl proxy = new ProxyImpl();
+
+        // host name used to create proxy connect request
+        // after creating the socket to proxy
+        final String hostName = event.getConnection().getHostname();
+        final ProxyHandler proxyHandler = new ProxyHandlerImpl();
+        final Map<String, String> proxyHeader = getAuthorizationHeader();
+        proxy.configure(hostName, proxyHeader, proxyHandler, transport);
+
+        transport.addTransportLayer(proxy);
+
+        if (TRACE_LOGGER.isInfoEnabled()) {
+            TRACE_LOGGER.info("addProxyHandshake: hostname[" + hostName +"]");
+        }
+    }
+
+    @Override
+    protected void notifyTransportErrors(final Event event) {
+        final Transport transport = event.getTransport();
+        final Connection connection = event.getConnection();
+        if (connection == null || transport == null) {
+            return;
+        }
+
+        final ErrorCondition errorCondition = transport.getCondition();
+        final String hostName = event.getReactor().getConnectionAddress(connection);
+        final ProxySelector proxySelector = ProxySelector.getDefault();
+
+        if (errorCondition == null
+                || !(errorCondition.getCondition().equals(ConnectionError.FRAMING_ERROR)
+                        || errorCondition.getCondition().equals(AmqpErrorCode.PROTON_IO_ERROR))
+                || proxySelector == null
+                || StringUtil.isNullOrEmpty(hostName)) {
+            return;
+        }
+
+        final String[] hostNameParts = hostName.split(":");
```

Comment:
When this error happens, we should have a valid host name. So, not sure whether we need this.